grpc_asyncio/socket.pyx

```diff
@@ -80,19 +84,24 @@ cdef class Socket:
     cdef void connect(self, object host, object port, grpc_custom_connect_callback g_connect_cb):
         assert not self.task_connect
 
+        self.g_connect_cb = g_connect_cb
         self.task_connect = asyncio.create_task(
-            asyncio.open_connection(host, port)
+            open_connection(host, port)
         )
         self.task_connect.add_done_callback(self._connect_cb)
-        self.g_connect_cb = g_connect_cb
 
     cdef void read(self, char * buffer_, size_t length, grpc_custom_read_callback g_read_cb):
         assert not self.task_read
 
-        self.task_read = asyncio.create_task(
-            self.reader.read(n=length)
-        )
-        self.task_read.add_done_callback(self._read_cb)
+        if self.reader.is_data_available(length):
+          self._read_cb(self.reader.read(n=length))
```

Comment:
This is a `coro` and a coro does not implement the Future interface, indeed the coro won't be scheduled in that scenario, not even started.

My suggestion is to implement a new method in the `StreamReader` for reading without having to wait and then if you want to wrap the result in a `Future` this will allow you to reuse the `_read_cb` function, for example

```
data = self.reader.read_nowait(n)
if data:
    f = Future()
    f.set_result(data)
    self._read_cb(f)
else:
    # awaitable version of the code
```